### ISSUE TITLE
Inductor backend exception (not a graph break planned by dynamo) on aten.nonzero.default from boolean-mask assignment under torch.compile(dynamic=True)

### ISSUE THREAD

**Issue by t0mnt** [pytorch/pytorch#194166]
### 🐛 Describe the bug

Human rephrasing of Claude summary.

Compiling a module that applies BatchNorm to a boolean-masked subset and writes it back (out[mask] = norm(h[mask])) fails in the inductor backend with Exception: aten.nonzero.default, then falls back via a graph break. Dynamo itself handles the data-dependent nonzero (unbacked symints); it is the backend that throws, and the error text itself says "Hint: Report an issue to the backend compiler repo". Should result in either a lowering for the masked pattern or a graph break planned by dynamo, not a backend exception on a documented-supported op.

Ablation: backend="inductor" (default mode, dynamic=True) throws; eager is fine.

Minimal reproduction:

```
import torch

class MaskedBN(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.norm = torch.nn.BatchNorm1d(8)

    def forward(self, h, mask_bool):
        out = h.clone()
        out[mask_bool] = self.norm(h[mask_bool])
        return out

m = MaskedBN().cuda()
f = torch.compile(m, dynamic=True)
h = torch.randn(64, 8, device="cuda", requires_grad=True)
mask = torch.rand(64, device="cuda") > 0.3
f(h, mask).sum().backward()

```
Observed in a real training run (particle-physics GNN, per-step hot path) and reduced to the above. Severity: functional fallback since training proceeds through the break. A performance bug, but the break lands inside the per-step region and the exception path is undocumented.

### Error logs

W0818 21:05:28 torch/_dynamo/exc.py:518 [4/0] Backend compiler exception
  Explanation: Backend compiler `inductor` failed with aten.nonzero.default. Adding a graph break.
  Hint: Report an issue to the backend compiler repo.
  Developer debug context: Backend: inductor
    Exception:aten.nonzero.default
    Traceback:
      File ".../plaingraphgps.py", line 192, in torch_dynamo_resume_in_forward_at_191
        return out
/usr/local/lib/python3.12/dist-packages/torch/autograd/graph.py:829: UserWarning:
  Error detected in IndexPutBackward0. Traceback of forward call that caused the error:
  File ".../plaingraphgps.py", line 191, in torch_dynamo_resume_in_forward_at_191
    out[mask_bool] = self.norm(h[mask_bool])   # BatchNorm over the real nodes only

### Versions

2.8.0a0+34c6371d24.nv25.08

cc @chauhang @penguinwu @ezyang @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo
